syntax = "proto3";

package emergency;

// Service de gestion des urgences
service EmergencyService {
  // Récupérer un véhicule par ID
  rpc GetVehicle (VehicleRequest) returns (Vehicle);
  
  // Récupérer tous les véhicules
  rpc GetAllVehicles (Empty) returns (VehicleList);
  
  // Récupérer les véhicules disponibles par type
  rpc GetAvailableVehicles (VehicleTypeRequest) returns (VehicleList);
  
  // Créer un nouveau véhicule
  rpc CreateVehicle (VehicleInput) returns (Vehicle);
  
  // Mettre à jour le statut d'un véhicule
  rpc UpdateVehicleStatus (StatusUpdate) returns (Vehicle);
  
  // Créer une intervention
  rpc CreateIntervention (InterventionInput) returns (Intervention);
  
  // Récupérer les interventions actives
  rpc GetActiveInterventions (Empty) returns (InterventionList);
  
  // Terminer une intervention
  rpc CompleteIntervention (InterventionRequest) returns (Intervention);
}

// Messages
message Empty {}

message VehicleRequest {
  int32 id = 1;
}

message VehicleTypeRequest {
  string vehicle_type = 1;
}

message Vehicle {
  int32 id = 1;
  string vehicle_type = 2;  // ambulance, fire_truck, police_car
  string identifier = 3;     // Numéro du véhicule (ex: AMB-001)
  string status = 4;         // available, on_mission, maintenance
  double latitude = 5;
  double longitude = 6;
  string station = 7;        // Caserne/Station de rattachement
  int32 crew_size = 8;       // Nombre de personnes à bord
  string created_at = 9;
}

message VehicleInput {
  string vehicle_type = 1;
  string identifier = 2;
  string status = 3;
  double latitude = 4;
  double longitude = 5;
  string station = 6;
  int32 crew_size = 7;
}

message VehicleList {
  repeated Vehicle vehicles = 1;
}

message StatusUpdate {
  int32 vehicle_id = 1;
  string new_status = 2;
  double latitude = 3;
  double longitude = 4;
}

message InterventionRequest {
  int32 id = 1;
}

message Intervention {
  int32 id = 1;
  string intervention_type = 2;  // medical, fire, crime, accident
  string priority = 3;            // low, medium, high, critical
  string address = 4;
  double latitude = 5;
  double longitude = 6;
  string status = 7;              // pending, in_progress, completed
  int32 assigned_vehicle_id = 8;
  string description = 9;
  string created_at = 10;
  string completed_at = 11;
}

message InterventionInput {
  string intervention_type = 1;
  string priority = 2;
  string address = 3;
  double latitude = 4;
  double longitude = 5;
  int32 assigned_vehicle_id = 6;
  string description = 7;
}

message InterventionList {
  repeated Intervention interventions = 1;
}
